docs/snackbar.js

```diff
@@ -0,0 +1,77 @@
+export {showSnackbar};
+
+const ANIMATION_DURATION_MS = 500;
+const SNACKBAR_DURATION_MS = 3000;
+
+// Track number of messages shown so we don't accidentally hide a later message.
+let showNumber = 0;
+
+let snackbarOriginalClassName;
+let iconOriginalClassName;
+
+/**
+ * Shows a snackbar notification.
+ *
+ * @param {string} msg the message to be shown
+ * @param {string[]} snackbarClasses classes to be added to the snackbar
+ * @param {string[]} iconClasses classes to be added to the icon
+ * @param {number} duration Number of milliseconds to show the message. If
+ *     non-positive, message stays up until this method is called again.
+ */
+function showSnackbar(
+    msg, snackbarClasses, iconClasses, duration = SNACKBAR_DURATION_MS) {
+  let snackbar = document.getElementById('snackbar');
+
+  // Lazily initialize the snackbar.
+  if (!snackbar) {
+    initializeSnackbar();
+    snackbar = document.getElementById('snackbar');
+  }
+
+  if (snackbarClasses) {
+    snackbarClasses.forEach(
+        (snackbarClass) => snackbar.classList.add(snackbarClass));
+  }
+
+  const icon = document.getElementById('snackbar-icon');
+  if (iconClasses) {
+    iconClasses.forEach((iconClass) => icon.classList.add(iconClass));
+  }
+
+  document.getElementById('snackbar-text').innerText = msg;
+
+  $('#snackbar').fadeIn(ANIMATION_DURATION_MS);
+
+  const currentShow = ++showNumber;
+
+  if (duration > 0) {
+    setTimeout(() => {
+      // Only clear the snackbar if the current shown snackbar is the latest.
+      if (currentShow === showNumber) {
+        showNumber = 0;  // reset the counter to prevent overflow
+        $('#snackbar').fadeOut(ANIMATION_DURATION_MS, () => {
+          // In case another snackbar is already being shown again, we don't
+          // want to reset the styles.
+          if (showNumber > 0) return;
```

Comment:
What happens if fadeIn is called while fadeOut is in progress? Does it get totally canceled, and the callback isn't called? If so, then this guard isn't necessary, right?

If the callback is called even if another message has faded in, then I think this is a bug. What if message1 comes in, timeout hits, it starts fading out, then message2 comes in, timeout hits, it sets showNumber to 0, then starts fading out, then message1's fadeOut callback hits, and resets the style before message2's fadeOut finishes. I would suggest resetting the showNumber inside the fadeOut callback, or not even resetting it (9007199254740991 messages ought to be enough for anyone).

Reply:
I think we need to keep the `showNumber` reset because that's how we identify whether or not there is a snackbar already being shown again and bail out of resetting the styles.

Moved the reset to within the fade out callback just in case.

Reply:
I don't understand this now. You're not setting showNumber to 0 unless you get past this guard, so when will you ever pass this guard?

Can't you just check here that currentShow === showNumber like above? And not reset?

Reply:
Oops ... silly mistake.

Yup that works better.